### ISSUE TITLE
Refresh tokens for GitHub.

### ISSUE THREAD

**Issue by adam-azarchs** [git-ecosystem/git-credential-manager#789]
I'm not 100% sure, but as far as I can tell, at the moment the GitHub OAuth2 flow is generating long-lived tokens. It would be nice to have the ability to use [short-lived credentials + refresh tokens](https://docs.github.com/en/developers/apps/building-github-apps/refreshing-user-to-server-access-tokens).  That flow appears to already be mandatory for BitBucket; it would be nice to have that as at least an option for GitHub as well.

Short-lived oauth tokens with refresh tokens are more secure than long-lived tokens because if your token gets exfiltrated then it'll expire quickly, and if the attacker uses the refresh token then you should at least be able to detect that quickly.  I'm sure long-lived tokens are fine for many situations, but for organizations which are more paranoid, or where the risk of exposure is higher (e.g. situations where developers are logging in from a shared machine) it would be helpful to have the ability to enforce that higher level of security.

Implementation-wise this should be pretty straightforward, since it's already supported for BitBucket and GitLab, though it might require making a separate client secret for GCM in this mode, since it looks from the docs like GitHub's configuration on whether oauth credentials are short-lived or not is on a per-application basis.

**Comment by hickford** [comment]
Good question. [Might access token expiry even have prevented or reduced the 2022 stolen OAuth token attack?](
https://security.stackexchange.com/questions/263314/would-refresh-tokens-have-prevented-the-github-stolen-oauth-token-attack) FWIW [GitHub *Apps*](https://docs.github.com/en/developers/apps/getting-started-with-apps/differences-between-github-apps-and-oauth-apps#token-based-identification) offer expiring access tokens [as an optional feature](https://docs.github.com/en/developers/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps).  Why not offer expiring access tokens for OAuth too?

Interestingly Git*Lab* recently [removed non-expiring access tokens in version 15.0](https://docs.gitlab.com/ee/update/removals.html#oauth-tokens-without-an-expiration). This is breaking change for apps which don't support refresh flow -- users would have to reauthenticate every 2 hours.

**Comment by mjcheetham** [comment]
This issue requires migration to the GitHub App infrastructure, which has several issues when using a public client app rather than a confidential client (specifically around org installation and consent).

Blocked on GitHub platform enhancements.
